### DISCUSSION TITLE
Support the codetools package

### DISCUSSION THREAD

**Discussion by llrs** [r-devel/r-project-sprint#11] [Ideas]
[codetools](https://cran.r-project.org/package=codetools) is used internally by R to [check code of packages](https://github.com/r-devel/r-svn/blob/master/src/library/tools/R/QC.R). In addition, codetools is used by several other packages that analyze code. I suggest this because I use the targets package that depends on the behavior of codetools and I have found some quirks on how it works.

However, codetools itself has  some [improvements/bug fixes pending](https://gitlab.com/luke-tierney/codetools/-/issues). Some of them are open since a couple years ago.

It would be nice if some of the issues could be reviewed and maybe with a patch proposal. These issues might have a high impact in packages:
- [identification of unassigned expressions](https://gitlab.com/luke-tierney/codetools/-/issues/14)
- [Incorrectly warning about unused arguments](https://gitlab.com/luke-tierney/codetools/-/issues/12)
- [ignores objects loaded by data()](https://gitlab.com/luke-tierney/codetools/-/issues/8)
- [Missing some variables from the global/parent? environment](https://gitlab.com/luke-tierney/codetools/-/issues/3)
- [fails to identify globals in functions produced by a 'factory' pattern](https://gitlab.com/luke-tierney/codetools/-/issues/1)

The code is generated from a [long file](https://gitlab.com/luke-tierney/codetools/-/blob/master/noweb/codetools.nw) using make and makefiles. 
It also has a [test file](https://gitlab.com/luke-tierney/codetools/-/blob/master/tests/tests.R) that could be expanded to ensure bugs fixed remain fixed.

**Comment by HenrikBengtsson** [comment]
> The code is generated from a long file using make and makefiles.

It's pretty cool that **codetools** is developed using the literate-programming principles by Donald Knuth.

**Comment by mmaechler** [comment]
codetools has a unique author/maintainer.  I think you should *first* ask about his thoughts in private ..

**Reply by llrs** [reply]
Thanks for the feedback Martin. I knew this is quite on limit to contribute to R but I wanted to gauge if there was interest on it (I won't be on the sprint). 

Given the current relative high number of upvotes (at this moment) I'll reach out the maintainer as I am affected by its current behavior through a package that depends on it (targets).

**Comment by llrs** [comment]
I contacted the maintainer. He did close some issues and commented on others.
